### ISSUE TITLE
component is not working in home-assistant 0.61

### ISSUE THREAD

**Issue by lnxusr** [nugget/python-anthemav#6]
There is no error in the ui of home assistant. There is no Anthem card visible after restart. I just installed home assistant for the first time. I have other components working: Google Cast, Kodi, fritz box.

The configuration syntax is correct and accepted by home assistant.

Version info:
```
bash-4.4# pip3 show anthemav
Name: anthemav
Version: 1.1.8
Summary: Python API for controlling Anthem Receivers
Home-page: https://github.com/nugget/python-anthemav
Author: David McNett
Author-email: nugget@macnugget.org
License: LICENSE
Location: /usr/lib/python3.6/site-packages
Requires: 
```

The logs show:
```
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] refresh_all
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Zone 1 Power (Z1POW) -> On (1)
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Front Panel Brightness (FPB) -> High (3)
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Zone 1 Volume (Z1VOL) -> -38
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Region (IDR) -> EU 
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Model (IDM) -> MRX 310 
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Software version (IDS) -> 1.2.9
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Software build date (IDB) -> Apr 28 2015
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Hardware version (IDH) -> 1
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: MAC address (IDN) -> 7CB77B010BA8
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Tx status (ECH) -> On (1)
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Standby IP control (SIP) -> On (1)
2018-01-16 13:30:11 WARNING (MainThread) [anthemav.protocol] Invalid command: ICN?
2018-01-16 13:30:11 WARNING (MainThread) [anthemav.protocol] Invalid command: Z1INP?
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Zone 1 mute (Z1MUT) -> Unmuted (0)
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Zone 1 ARC (Z1ARC) -> On (1)
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Video input resolution (Z1VIR) -> 1080p24 (4)
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Active horizontal video resolution (pixels) (Z1IRH) -> 1920
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Active vertical video resolution (pixels) (Z1IRV) -> 1080
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Audio input channels (Z1AIC) -> 5.1 channel (4)
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Audio input format (Z1AIF) -> PCM (2)
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Audio input bitrate (kbps) (Z1BRT) -> 3840
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Audio input sampling rate (hKz) (Z1SRT) -> 48
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Audio input name (Z1AIN) -> Multi PCM
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Audio input rate name (Z1AIR) -> 48kbps
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Audio listening mode (Z1ALM) -> None (00)
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Dolby digital dynamic range (Z1DYN) -> Normal (0)
2018-01-16 13:30:11 INFO (MainThread) [anthemav.protocol] Unchanged: Dolby digital dialog normalization (dB) (Z1DIA) -> 0
```

When running the anthem-monitor inside the homeassistant docker container I get:

```
INFO:anthemav.connection:Connecting to Anthem AVR at 10.10.10.106:14999
INFO:anthemav.protocol:Connection established to AVR
DEBUG:anthemav.protocol:Write buffer limits 16384 to 65536
DEBUG:anthemav.protocol:> b'ECH1;'
INFO:anthemav.protocol:Sending out mass query for all attributes
DEBUG:anthemav.protocol:> b'Z1POW?;'
DEBUG:anthemav.protocol:> b'IDM?;'
DEBUG:anthemav.protocol:Received 1 bytes from AVR: ;
DEBUG:anthemav.protocol:Received 7 bytes from AVR: Z1POW1;
DEBUG:anthemav.protocol:assembled message Z1POW1
INFO:anthemav.protocol:Unchanged: Zone 1 Power (Z1POW) -> On (1)
DEBUG:anthemav.protocol:no new data encountered
DEBUG:anthemav.protocol:Received 12 bytes from AVR: IDMMRX 310 ;
DEBUG:anthemav.protocol:assembled message IDMMRX 310 
INFO:anthemav.protocol:Unchanged: Model (IDM) -> MRX 310 
DEBUG:anthemav.protocol:no new data encountered
INFO:anthemav.protocol:refresh_all
DEBUG:anthemav.protocol:> b'Z1POW?;'
DEBUG:anthemav.protocol:> b'FPB?;'
DEBUG:anthemav.protocol:> b'Z1VOL?;'
DEBUG:anthemav.protocol:> b'IDR?;'
DEBUG:anthemav.protocol:> b'IDM?;'
DEBUG:anthemav.protocol:> b'IDS?;'
DEBUG:anthemav.protocol:> b'IDB?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'IDH?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'IDN?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'ECH?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'SIP?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'ICN?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1INP?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1MUT?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1ARC?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1VIR?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1IRH?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1IRV?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1AIC?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1AIF?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1BRT?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1SRT?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1AIN?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1AIR?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1ALM?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1DYN?;'
WARNING:asyncio:socket.send() raised exception.
DEBUG:anthemav.protocol:> b'Z1DIA?;'
WARNING:asyncio:socket.send() raised exception.
WARNING:anthemav.protocol:Lost connection to receiver: [Errno 104] Connection reset by peer
```

**Comment by lnxusr** [comment]
edit: the old errors still occur after a restart of home-assistant.

I just updated to home assistant 0.61.1. The error has changed:

```
2018-01-18 12:55:48 ERROR (MainThread) [homeassistant.components.media_player] Error while setting up platform anthemav
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/homeassistant/helpers/entity_component.py", line 171, in _async_setup_platform
    SLOW_SETUP_MAX_WAIT, loop=self.hass.loop)
  File "/usr/lib/python3.6/asyncio/tasks.py", line 358, in wait_for
    return fut.result()
  File "/usr/lib/python3.6/asyncio/futures.py", line 245, in result
    raise self._exception
  File "/usr/lib/python3.6/asyncio/tasks.py", line 180, in _step
    result = coro.send(None)
  File "/usr/lib/python3.6/site-packages/homeassistant/components/media_player/anthemav.py", line 63, in async_setup_platform
    _LOGGER.debug("dump_rawdata: %s", avr.protocol.dump_rawdata)
  File "/usr/lib/python3.6/site-packages/anthemav/protocol.py", line 855, in dump_rawdata
    attrs = vars(self.transport)
TypeError: vars() argument must have __dict__ attribute
```
anthemav_monitor no longer throws an error. The component still doesn't show up in home assistant.

**Comment by lnxusr** [comment]
I've commented the lines referring to ICN and Z1INP in the anthemav python site-package. This make the error disappear but home assistant keeps logging a lot of anthemav data. No media_player device is added to the ui.

**Comment by lnxusr** [comment]
I should have mentioned I own a mrx-310.

**Comment by petonic** [comment]
I've got the same problem, but I don't think that this library is at fault.  I closed this repo, and it's running just fine against my Anthem MRX 710.... (need to write a small wrapper that calls monitor()):

```
python3 tools.py --host 192.168.86.54 --port 14999                                                                                               (master✱)
INFO:__main__:Connecting to Anthem AVR at 192.168.86.54:14999
INFO:anthemav.connection:Connecting to Anthem AVR at 192.168.86.54:14999
INFO:anthemav.protocol:Connection established to AVR
INFO:anthemav.protocol:Sending out mass query for all attributes
INFO:__main__:Power state is False
INFO:__main__:Power state is False
INFO:anthemav.protocol:New Value: Zone 1 Power (Z1POW) -> On (1)
INFO:anthemav.protocol:Power on detected, refreshing all attributes
INFO:anthemav.protocol:New Value: Model (IDM) -> MRX 710
INFO:__main__:Callback invoked: Z1POW1
INFO:__main__:Callback invoked: IDMMRX 710
INFO:anthemav.protocol:refresh_all
INFO:anthemav.protocol:New Value: Front Panel Brightness (FPB) -> Medium (2)
INFO:anthemav.protocol:New Value: Audio input format (Z1AIF) -> No audio (0)
INFO:anthemav.protocol:New Value: MAC address (IDN) -> 7CB77B011741
INFO:anthemav.protocol:New Value: Hardware version (IDH) -> 1
INFO:anthemav.protocol:New Value: Zone 1 current input (Z1INP) -> 1
INFO:anthemav.protocol:New Value: Dolby digital dynamic range (Z1DYN) -> Normal (0)
INFO:anthemav.protocol:New Value: Dolby digital dialog normalization (dB) (Z1DIA) -> 0
INFO:anthemav.protocol:New Value: Zone 1 mute (Z1MUT) -> Unmuted (0)
INFO:anthemav.protocol:New Value: Audio input bitrate (kbps) (Z1BRT) -> 1536
INFO:anthemav.protocol:New Value: Audio input channels (Z1AIC) -> No audio (0)
INFO:anthemav.protocol:New Value: Zone 1 Volume (Z1VOL) -> -35
INFO:anthemav.protocol:New Value: Software build date (IDB) -> Apr 28 2015
INFO:anthemav.protocol:New Value: Standby IP control (SIP) -> On (1)
INFO:anthemav.protocol:New Value: Audio input sampling rate (hKz) (Z1SRT) -> 0
INFO:anthemav.protocol:New Value: Tx status (ECH) -> On (1)
INFO:anthemav.protocol:New Value: Active horizontal video resolution (pixels) (Z1IRH) -> 720
INFO:anthemav.protocol:New Value: Region (IDR) -> US
INFO:anthemav.protocol:New Value: Active vertical video resolution (pixels) (Z1IRV) -> 480
INFO:anthemav.protocol:New Value: Zone 1 ARC (Z1ARC) -> Off (0)
INFO:anthemav.protocol:Unchanged: Zone 1 Power (Z1POW) -> On (1)
INFO:anthemav.protocol:New Value: Audio input rate name (Z1AIR) -> 0kHz
INFO:anthemav.protocol:New Value: Audio input name (Z1AIN) -> ?No Input
INFO:anthemav.protocol:Unchanged: Model (IDM) -> MRX 710
INFO:anthemav.protocol:New Value: Audio listening mode (Z1ALM) -> None (00)
INFO:anthemav.protocol:New Value: Video input resolution (Z1VIR) -> No video (0)
INFO:anthemav.protocol:New Value: Software version (IDS) -> 1.2.9
INFO:__main__:Callback invoked: FPB2
INFO:__main__:Callback invoked: Z1AIF0
INFO:__main__:Callback invoked: IDN7CB77B011741
INFO:__main__:Callback invoked: IDH1
INFO:__main__:Callback invoked: Z1INP1
INFO:__main__:Callback invoked: Z1DYN0
INFO:__main__:Callback invoked: Z1DIA0
INFO:__main__:Callback invoked: Z1MUT0
INFO:__main__:Callback invoked: Z1BRT1536
INFO:__main__:Callback invoked: Z1AIC0
INFO:__main__:Callback invoked: Z1VOL-35
INFO:__main__:Callback invoked: IDBApr 28 2015
INFO:__main__:Callback invoked: SIP1
INFO:__main__:Callback invoked: Z1SRT0
INFO:__main__:Callback invoked: ECH1
INFO:__main__:Callback invoked: Z1IRH720
INFO:__main__:Callback invoked: IDRUS
INFO:__main__:Callback invoked: Z1IRV480
INFO:__main__:Callback invoked: Z1ARC0
INFO:__main__:Callback invoked: Z1AIR0kHz
INFO:__main__:Callback invoked: Z1AIN?No Input
INFO:__main__:Callback invoked: Z1ALM00
INFO:__main__:Callback invoked: Z1VIR0
INFO:__main__:Callback invoked: IDS1.2.9
INFO:anthemav.protocol:New Value: Active input count (ICN) -> 5
WARNING:anthemav.protocol:ICN update received
INFO:__main__:Callback invoked: ICN5
INFO:anthemav.protocol:New Value: Input 1 is called HDMI 1
INFO:anthemav.protocol:New Value: Input 2 is called HDMI 2
INFO:anthemav.protocol:New Value: Input 3 is called HDMI 3
INFO:__main__:Callback invoked: ISN01HDMI 1
INFO:__main__:Callback invoked: ISN02HDMI 2
INFO:__main__:Callback invoked: ISN03HDMI 3
INFO:anthemav.protocol:New Value: Input 4 is called FM
INFO:anthemav.protocol:New Value: Input 5 is called AM
INFO:__main__:Callback invoked: ISN04FM
INFO:__main__:Callback invoked: ISN05AM
```

**Comment by wvh-github** [comment]
@petonic 
Hi Mike, 

I don't think the library is at fault either, but I have no idea what's going on and nobody is responding to the home-assistant bug. The home-assistant component is also created by David (nugget) and I was hoping to get his attention here.

Looking at the current state I think the component should be removed from Home-Assistant.

**Comment by achatham** [comment]
I tracked this down. Fixing it in python-anthemav is probably best.

The issues is home assistant injects uvloop into asyncio. It's enabled in homeassistant/__main__.py 

```
import uvloop
asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())
```

If you add that code and use python-anthemav it will crash in the same place, because uvloop uses a PythonC transport object, while default asyncio uses a python object. `vars()` only works for pure python. In anthemav/protocol.py `dump_rawdata` calls `vars(self.transport)`.

The fix is probably to just not run that line or put a try/except around it.

**Comment by nugget** [comment]
Resolved by #9 and also paired with corresponding home assistant PR 
https://github.com/home-assistant/home-assistant/pull/21273

Thanks!

**Comment by achatham** [comment]
BTW, I deleted the offending line in home assistant, so it's not an issue in practice. But still probably worth fixing here. This was hard to figure out!
